quickwit/quickwit-ingest-api/src/doc_batch.rs

```diff
@@ -0,0 +1,420 @@
+//
+// Quickwit is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at hello@quickwit.io.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use bytes::buf::Writer;
+use bytes::{Buf, BufMut, Bytes, BytesMut};
+use serde::Serialize;
+
+use crate::metrics::INGEST_METRICS;
+use crate::DocBatch;
+
+#[derive(Debug)]
+/// Represents a command that can be stored in doc batch.
+pub enum DocCommand<T>
+where T: Buf
+{
+    Ingest { payload: T },
+    Commit,
+    // ... more to come?
+}
+
+/// We can use this byte to track both commands and their version changes
+/// If serialization protocol changes, we can just use the next number
+const INGEST_CODE_V1: u8 = 0;
+const COMMIT_CODE_V1: u8 = 1;
+
+impl<T> DocCommand<T>
+where T: Buf
+{
+    /// Returns the
+    pub fn code(&self) -> u8 {
+        match self {
+            DocCommand::Ingest { payload: _ } => INGEST_CODE_V1,
+            DocCommand::Commit => COMMIT_CODE_V1,
+        }
+    }
+
+    /// Builds a command for bytes::Buf
+    pub fn read(buf: T) -> Self {
+        let mut buf = buf;
+        match buf.get_u8() {
+            INGEST_CODE_V1 => DocCommand::Ingest { payload: buf },
+            COMMIT_CODE_V1 => DocCommand::Commit,
+            other => panic!("Encountered unknown command: code {other}"),
+        }
+    }
+
+    /// Copies the command to the end of bytes::BufMut while returning the number of bytes copied
+    pub fn write(self, buf: &mut impl BufMut) -> usize {
+        buf.put_u8(self.code());
+        match self {
+            DocCommand::Ingest { payload } => {
+                let len = payload.remaining();
+                buf.put(payload);
+                len + 1
+            }
+            DocCommand::Commit => 1,
+        }
+    }
+}
+
+/// Builds DocBatch from individual commands
+pub struct DocBatchBuilder {
+    index_id: String,
+    concat_docs: BytesMut,
+    doc_lens: Vec<u64>,
+}
+
+impl DocBatchBuilder {
+    /// Creates a new batch builder with the given index name
+    pub fn new(index_id: String) -> Self {
+        Self {
+            index_id,
+            concat_docs: BytesMut::new(),
+            doc_lens: Vec::new(),
+        }
+    }
+
+    /// Adds an ingest command to the batch
+    pub fn ingest_doc(&mut self, payload: impl Buf + Default) -> usize {
```

Comment:
why does it requires `Default`?

Reply:
Leftover from one of the iterations where it was needed. Don't need it anymore. Great catch!